// ra-supabase-ui-materialui/esm/guessers/CreateGuesser.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import * as React from 'react';
import { useAPISchema } from 'ra-supabase-core';
import { CreateBase, useResourceContext } from 'ra-core';
import { CreateView, Loading } from 'ra-ui-materialui';
import { capitalize, singularize } from 'inflection';
import { inferElementFromType } from './inferElementFromType';
import { InferredElement } from './InferredElement';
import { editFieldTypes } from './editFieldTypes';
export var CreateGuesser = function (props) {
    var mutationOptions = props.mutationOptions, resource = props.resource, record = props.record, transform = props.transform, redirect = props.redirect, disableAuthentication = props.disableAuthentication, rest = __rest(props, ["mutationOptions", "resource", "record", "transform", "redirect", "disableAuthentication"]);
    return (React.createElement(CreateBase, { resource: resource, record: record, redirect: redirect, transform: transform, mutationOptions: mutationOptions, disableAuthentication: disableAuthentication },
        React.createElement(CreateGuesserView, __assign({}, rest))));
};
export var CreateGuesserView = function (props) {
    var _a = useAPISchema(), schema = _a.data, error = _a.error, isPending = _a.isPending;
    var resource = useResourceContext();
    var _b = React.useState(null), child = _b[0], setChild = _b[1];
    if (!resource) {
        throw new Error('CreateGuesser must be used withing a ResourceContext');
    }
    var _c = props.enableLog, enableLog = _c === void 0 ? process.env.NODE_ENV === 'development' : _c, rest = __rest(props, ["enableLog"]);
    React.useEffect(function () {
        var _a;
        if (isPending || error) {
            return;
        }
        var resourceDefinition = (_a = schema.definitions) === null || _a === void 0 ? void 0 : _a[resource];
        var requiredFields = (resourceDefinition === null || resourceDefinition === void 0 ? void 0 : resourceDefinition.required) || [];
        if (!resourceDefinition || !resourceDefinition.properties) {
            throw new Error("The resource ".concat(resource, " is not defined in the API schema"));
        }
        var inferredInputs = Object.keys(resourceDefinition.properties)
            .filter(function (source) { return source !== 'id'; })
            .filter(function (source) {
            return resourceDefinition.properties[source].format !== 'tsvector';
        })
            .map(function (source) {
            return inferElementFromType({
                name: source,
                types: editFieldTypes,
                description: resourceDefinition.properties[source].description,
                format: resourceDefinition.properties[source].format,
                type: (resourceDefinition.properties &&
                    resourceDefinition.properties[source] &&
                    typeof resourceDefinition.properties[source].type ===
                        'string'
                    ? resourceDefinition.properties[source].type
                    : 'string'),
                requiredFields: requiredFields,
                schema: schema,
            });
        });
        var inferredForm = new InferredElement(editFieldTypes.form, null, inferredInputs);
        setChild(inferredForm.getElement());
        if (!enableLog)
            return;
        var representation = inferredForm.getRepresentation();
        var components = ['Create']
            .concat(Array.from(new Set(Array.from(representation.matchAll(/<([^/\s>]+)/g))
            .map(function (match) { return match[1]; })
            .filter(function (component) { return component !== 'span'; }))))
            .sort();
        var warnings = inferredInputs
            .map(function (inferredInput) { return inferredInput.getWarning(); })
            .filter(function (warning) { return warning != null; });
        // eslint-disable-next-line no-console
        console.log("Guessed Create:\n            \nimport { ".concat(components.join(', '), " } from 'react-admin';\n            \nexport const ").concat(capitalize(singularize(resource)), "Create = () => (\n    <Create>\n").concat(representation, "\n    </Create>\n);").concat(warnings.length > 0
            ? warnings.map(function (warning) { return "\n\n".concat(warning); }).join('')
            : ''));
    }, [resource, isPending, error, schema, enableLog]);
    if (isPending)
        return React.createElement(Loading, null);
    if (error)
        return React.createElement("p", null,
            "Error: ",
            error.message);
    return React.createElement(CreateView, __assign({}, rest), child);
};
