// ra-supabase-ui-materialui/esm/guessers/ListGuesser.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import * as React from 'react';
import { useAPISchema } from 'ra-supabase-core';
import { ListBase, useResourceContext } from 'ra-core';
import { listFieldTypes, ListView, Loading, SearchInput, } from 'ra-ui-materialui';
import { capitalize, singularize } from 'inflection';
import { inferElementFromType } from './inferElementFromType';
import { InferredElement } from './InferredElement';
import { editFieldTypes } from './editFieldTypes';
export var ListGuesser = function (props) {
    var debounce = props.debounce, disableAuthentication = props.disableAuthentication, disableSyncWithLocation = props.disableSyncWithLocation, exporter = props.exporter, filter = props.filter, filterDefaultValues = props.filterDefaultValues, perPage = props.perPage, queryOptions = props.queryOptions, resource = props.resource, sort = props.sort, storeKey = props.storeKey, rest = __rest(props, ["debounce", "disableAuthentication", "disableSyncWithLocation", "exporter", "filter", "filterDefaultValues", "perPage", "queryOptions", "resource", "sort", "storeKey"]);
    return (React.createElement(ListBase, { debounce: debounce, disableAuthentication: disableAuthentication, disableSyncWithLocation: disableSyncWithLocation, exporter: exporter, filter: filter, filterDefaultValues: filterDefaultValues, perPage: perPage, queryOptions: queryOptions, resource: resource, sort: sort, storeKey: storeKey },
        React.createElement(ListGuesserView, __assign({}, rest))));
};
export var ListGuesserView = function (props) {
    var _a = useAPISchema(), schema = _a.data, error = _a.error, isPending = _a.isPending;
    var resource = useResourceContext();
    var _b = React.useState(null), child = _b[0], setChild = _b[1];
    var _c = React.useState(undefined), filters = _c[0], setFilters = _c[1];
    if (!resource) {
        throw new Error('ListGuesser must be used withing a ResourceContext');
    }
    var _d = props.enableLog, enableLog = _d === void 0 ? process.env.NODE_ENV === 'development' : _d, rest = __rest(props, ["enableLog"]);
    React.useEffect(function () {
        var _a, _b, _c;
        if (isPending || error) {
            return;
        }
        var resourceDefinition = (_a = schema.definitions) === null || _a === void 0 ? void 0 : _a[resource];
        if (!resourceDefinition || !resourceDefinition.properties) {
            throw new Error("The resource ".concat(resource, " is not defined in the API schema"));
        }
        var inferredFields = Object.keys(resourceDefinition.properties)
            .filter(function (source) {
            return resourceDefinition.properties[source].format !== 'tsvector';
        })
            .map(function (source) {
            return inferElementFromType({
                name: source,
                types: listFieldTypes,
                description: resourceDefinition.properties[source].description,
                format: resourceDefinition.properties[source].format,
                type: (resourceDefinition.properties &&
                    resourceDefinition.properties[source] &&
                    typeof resourceDefinition.properties[source].type ===
                        'string'
                    ? resourceDefinition.properties[source].type
                    : 'string'),
                schema: schema,
            });
        });
        var inferredTable = new InferredElement(listFieldTypes.table, null, inferredFields);
        setChild(inferredTable.getElement());
        var rowFilters = (_c = (_b = schema
            .paths["/".concat(resource)].get.parameters) === null || _b === void 0 ? void 0 : _b.filter(function (obj) {
            return obj['$ref'].includes('rowFilter');
        }).map(function (obj) { return obj['$ref'].split('.').pop(); })) !== null && _c !== void 0 ? _c : [];
        var inferredInputsForFilters = rowFilters
            .filter(function (source) {
            return resourceDefinition.properties[source].format !== 'tsvector';
        })
            .map(function (source) {
            var field = resourceDefinition.properties[source];
            return inferElementFromType({
                name: source,
                types: editFieldTypes,
                description: field.description,
                format: field.format,
                type: field.type,
                schema: schema,
            });
        });
        if (rowFilters.some(function (source) {
            return resourceDefinition.properties[source].format === 'tsvector';
        })) {
            var fullTextSearchSource = rowFilters.find(function (source) {
                return resourceDefinition.properties[source].format === 'tsvector';
            });
            var field = resourceDefinition.properties[fullTextSearchSource];
            inferredInputsForFilters.unshift(inferElementFromType({
                name: "".concat(fullTextSearchSource, "@fts"),
                types: {
                    string: {
                        component: SearchInput,
                        representation: function (props) {
                            return "<SearchInput alwaysOn source=\"".concat(props.source, "\" />");
                        },
                    },
                },
                description: field.description,
                format: 'tsvector',
                props: {
                    alwaysOn: true,
                    parse: function (value) { return (value ? "".concat(value, ":*") : undefined); },
                    format: function (value) {
                        return value ? value.substring(0, value.length - 2) : '';
                    },
                },
                type: field.type,
                schema: schema,
            }));
        }
        if (inferredInputsForFilters.length > 0) {
            var filterElements = inferredInputsForFilters
                .map(function (inferredInput) { return inferredInput.getElement(); })
                .filter(function (el) { return el != null; });
            setFilters(filterElements);
        }
        if (!enableLog)
            return;
        var tableRepresentation = inferredTable.getRepresentation();
        var filterRepresentation = inferredInputsForFilters.length > 0
            ? "const filters = [\n".concat(inferredInputsForFilters
                .map(function (inferredInput) { return '    ' + inferredInput.getRepresentation(); })
                .join(',\n'), "\n];\n")
            : '';
        var fieldComponents = Array.from(tableRepresentation.matchAll(/<([^/\s>]+)/g))
            .map(function (match) { return match[1]; })
            .filter(function (component) { return component !== 'span'; });
        var filterComponents = Array.from(filterRepresentation.matchAll(/<([^/\s>]+)/g))
            .map(function (match) { return match[1]; })
            .filter(function (component) { return component !== 'span'; });
        var components = Array.from(new Set(__spreadArray(__spreadArray(['List'], fieldComponents, true), filterComponents, true)))
            .filter(function (component) {
            return component !== 'span' &&
                !component.match(/[A-Za-z]+\.[A-Za-z]+/i);
        })
            .sort();
        var warnings = inferredInputsForFilters
            .map(function (inferredInput) { return inferredInput.getWarning(); })
            .concat(inferredFields.map(function (field) { return field.getWarning(); }))
            .filter(function (warning) { return warning != null; });
        // eslint-disable-next-line no-console
        console.log("Guessed List:\n            \nimport { ".concat(components.join(', '), " } from 'react-admin';\n\n").concat(filterRepresentation, "\nexport const ").concat(capitalize(singularize(resource)), "List = () => (\n    <List").concat(filterRepresentation ? ' filters={filters}' : '', ">\n").concat(tableRepresentation, "\n    </List>\n);").concat(warnings.length > 0
            ? warnings.map(function (warning) { return "\n\n".concat(warning); }).join('')
            : ''));
    }, [resource, isPending, error, schema, enableLog]);
    if (isPending)
        return React.createElement(Loading, null);
    if (error)
        return React.createElement("p", null,
            "Error: ",
            error.message);
    return (React.createElement(ListView, __assign({ filters: filters }, rest), child));
};
