// ra-supabase-ui-materialui/esm/guessers/useCrudGuesser.js
import { useMemo } from 'react';
import { useAPISchema } from 'ra-supabase-core';
import { ListGuesser } from './ListGuesser';
import { CreateGuesser } from './CreateGuesser';
import { EditGuesser } from './EditGuesser';
import { ShowGuesser } from './ShowGuesser';
export var useCrudGuesser = function () {
    var _a = useAPISchema(), schema = _a.data, error = _a.error, isPending = _a.isPending;
    return useMemo(function () {
        if (isPending || error) {
            return [];
        }
        var resourceNames = Object.keys(schema.definitions);
        return resourceNames.map(function (name) {
            var _a;
            var resourcePaths = (_a = schema.paths["/".concat(name)]) !== null && _a !== void 0 ? _a : {};
            return {
                name: name,
                list: resourcePaths.get ? ListGuesser : undefined,
                show: resourcePaths.get ? ShowGuesser : undefined,
                edit: resourcePaths.patch ? EditGuesser : undefined,
                create: resourcePaths.post ? CreateGuesser : undefined,
            };
        });
    }, [schema, isPending, error]);
};
