// ra-supabase-ui-materialui/esm/ForgotPasswordForm.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import * as React from 'react';
import { CardActions, Stack, styled, Typography } from '@mui/material';
import { useResetPassword } from 'ra-supabase-core';
import { Form, required, useNotify, useTranslate } from 'ra-core';
import { Link, SaveButton, TextInput } from 'ra-ui-materialui';
/**
 * A component that renders a form for resetting the user password.
 */
export var ForgotPasswordForm = function () {
    var notify = useNotify();
    var translate = useTranslate();
    var _a = useResetPassword(), resetPassword = _a[1].mutateAsync;
    var submit = function (values) { return __awaiter(void 0, void 0, void 0, function () {
        var error_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, resetPassword({
                            email: values.email,
                        })];
                case 1:
                    _a.sent();
                    return [3 /*break*/, 3];
                case 2:
                    error_1 = _a.sent();
                    notify(typeof error_1 === 'string'
                        ? error_1
                        : typeof error_1 === 'undefined' || !error_1.message
                            ? 'ra.auth.sign_in_error'
                            : error_1.message, {
                        type: 'warning',
                        messageArgs: {
                            _: typeof error_1 === 'string'
                                ? error_1
                                : error_1 && error_1.message
                                    ? error_1.message
                                    : undefined,
                        },
                    });
                    return [3 /*break*/, 3];
                case 3: return [2 /*return*/];
            }
        });
    }); };
    return (React.createElement(Root, { onSubmit: submit },
        React.createElement("div", { className: SupabaseLoginFormClasses.container },
            React.createElement(Stack, { spacing: 1 },
                React.createElement(Typography, { variant: "h5", sx: {
                        textAlign: 'center',
                    } }, translate('ra-supabase.reset_password.forgot_password', { _: 'Forgot password?' })),
                React.createElement(Typography, { variant: "body2", color: "textSecondary", sx: {
                        textAlign: 'center',
                    } }, translate('ra-supabase.reset_password.forgot_password_details', {
                    _: 'Enter your email to receive a reset password link.',
                }))),
            React.createElement("div", { className: SupabaseLoginFormClasses.input },
                React.createElement(TextInput, { source: "email", label: translate('ra.auth.email', {
                        _: 'Email',
                    }), autoComplete: "email", fullWidth: true, validate: required() }))),
        React.createElement(CardActions, { sx: { flexDirection: 'column', gap: 1 } },
            React.createElement(SaveButton, { variant: "contained", type: "submit", className: SupabaseLoginFormClasses.button, label: translate('ra.action.reset_password', {
                    _: 'Reset password',
                }), icon: React.createElement(React.Fragment, null) }),
            React.createElement(Link, { to: "/login", variant: "body2" }, translate('ra-supabase.auth.back_to_login', {
                _: 'Back to login page',
            })))));
};
var PREFIX = 'RaSupabaseForgotPasswordForm';
var SupabaseLoginFormClasses = {
    container: "".concat(PREFIX, "-container"),
    input: "".concat(PREFIX, "-input"),
    button: "".concat(PREFIX, "-button"),
};
var Root = styled(Form, {
    name: PREFIX,
    overridesResolver: function (props, styles) { return styles.root; },
})(function (_a) {
    var _b;
    var theme = _a.theme;
    return (_b = {},
        _b["& .".concat(SupabaseLoginFormClasses.container)] = {
            padding: '0 1em 0 1em',
        },
        _b["& .".concat(SupabaseLoginFormClasses.input)] = {
            marginTop: '1em',
        },
        _b["& .".concat(SupabaseLoginFormClasses.button)] = {
            width: '100%',
        },
        _b);
});
