// ra-supabase-ui-materialui/esm/SetPasswordForm.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import * as React from 'react';
import { CardActions, styled, Typography } from '@mui/material';
import { Form, required, useNotify, useTranslate } from 'ra-core';
import { PasswordInput, SaveButton } from 'ra-ui-materialui';
import { useSetPassword, useSupabaseAccessToken } from 'ra-supabase-core';
/**
 * A component that renders a form for setting the current user password through Supabase.
 * Can be used for the first login after a user has been invited or to reset the password.
 */
export var SetPasswordForm = function () {
    var access_token = useSupabaseAccessToken();
    var refresh_token = useSupabaseAccessToken({
        parameterName: 'refresh_token',
    });
    var notify = useNotify();
    var translate = useTranslate();
    var _a = useSetPassword(), setPassword = _a[1].mutateAsync;
    var validate = function (values) {
        if (values.password !== values.confirmPassword) {
            return {
                password: 'ra-supabase.validation.password_mismatch',
                confirmPassword: 'ra-supabase.validation.password_mismatch',
            };
        }
        return {};
    };
    if (!access_token || !refresh_token) {
        if (process.env.NODE_ENV === 'development') {
            console.error('Missing access_token or refresh_token for set password');
        }
        return (React.createElement("div", { className: SupabaseLoginFormClasses.container },
            React.createElement("div", null, translate('ra-supabase.auth.missing_tokens'))));
    }
    var submit = function (values) { return __awaiter(void 0, void 0, void 0, function () {
        var error_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, setPassword({
                            access_token: access_token,
                            refresh_token: refresh_token,
                            password: values.password,
                        })];
                case 1:
                    _a.sent();
                    return [3 /*break*/, 3];
                case 2:
                    error_1 = _a.sent();
                    notify(typeof error_1 === 'string'
                        ? error_1
                        : typeof error_1 === 'undefined' || !error_1.message
                            ? 'ra.auth.sign_in_error'
                            : error_1.message, {
                        type: 'warning',
                        messageArgs: {
                            _: typeof error_1 === 'string'
                                ? error_1
                                : error_1 && error_1.message
                                    ? error_1.message
                                    : undefined,
                        },
                    });
                    return [3 /*break*/, 3];
                case 3: return [2 /*return*/];
            }
        });
    }); };
    return (React.createElement(Root, { onSubmit: submit, validate: validate },
        React.createElement("div", { className: SupabaseLoginFormClasses.container },
            React.createElement(Typography, { variant: "h5", gutterBottom: true, sx: {
                    textAlign: 'center',
                } }, translate('ra-supabase.set_password.new_password', {
                _: 'Choose your password',
            })),
            React.createElement("div", { className: SupabaseLoginFormClasses.input },
                React.createElement(PasswordInput, { source: "password", label: translate('ra.auth.password', {
                        _: 'Password',
                    }), autoComplete: "new-password", fullWidth: true, validate: required() })),
            React.createElement("div", null,
                React.createElement(PasswordInput, { source: "confirmPassword", label: translate('ra.auth.confirm_password', {
                        _: 'Confirm password',
                    }), fullWidth: true, validate: required() }))),
        React.createElement(CardActions, null,
            React.createElement(SaveButton, { variant: "contained", type: "submit", className: SupabaseLoginFormClasses.button, label: translate('ra.action.save') }))));
};
var PREFIX = 'RaSupabaseSetPasswordForm';
var SupabaseLoginFormClasses = {
    container: "".concat(PREFIX, "-container"),
    input: "".concat(PREFIX, "-input"),
    button: "".concat(PREFIX, "-button"),
};
var Root = styled(Form, {
    name: PREFIX,
    overridesResolver: function (props, styles) { return styles.root; },
})(function (_a) {
    var _b;
    var theme = _a.theme;
    return (_b = {},
        _b["& .".concat(SupabaseLoginFormClasses.container)] = {
            padding: '0 1em 0 1em',
        },
        _b["& .".concat(SupabaseLoginFormClasses.input)] = {
            marginTop: '1em',
        },
        _b["& .".concat(SupabaseLoginFormClasses.button)] = {
            width: '100%',
        },
        _b);
});
